src/type/validate.js

```diff
@@ -593,6 +759,143 @@ function createInputObjectCircularRefsValidator(
+
+  // This does a straight-forward DFS to find cycles.
+  // It does not terminate when a cycle was found but continues to explore
+  // the graph to find all possible cycles.
+  return function validateInputObjectDefaultValueCircularRefs(
+    inputObj: GraphQLInputObjectType,
+  ): void {
+    // Start with an empty object as a way to visit every field in this input
+    // object type and apply every default value.
+    return detectValueDefaultValueCycle(inputObj, {});
+  };
+
+  function detectValueDefaultValueCycle(
+    inputObj: GraphQLInputObjectType,
+    defaultValue: mixed,
+  ): void {
+    // If the value is a List, recursively check each entry for a cycle.
+    // Otherwise, only object values can contain a cycle.
+    if (isIterableObject(defaultValue)) {
+      for (const itemValue of defaultValue) {
+        detectValueDefaultValueCycle(inputObj, itemValue);
+      }
+      return;
+    } else if (!isObjectLike(defaultValue)) {
+      return;
+    }
+
+    // Check each defined field for a cycle.
+    for (const field of Object.values(inputObj.getFields())) {
+      const fieldType = getNamedType(field.type);
+
+      // Only input object type fields can result in a cycle.
+      if (!isInputObjectType(fieldType)) {
+        continue;
+      }
+
+      // If the provided value has this field defined, recursively check it for
+      // cycles instead of checking the field's default value.
+      if (hasOwnProperty(defaultValue, field.name)) {
+        detectValueDefaultValueCycle(fieldType, defaultValue[field.name]);
+        continue;
+      }
+
+      detectFieldDefaultValueCycle(inputObj, field, fieldType);
+    }
+  }
+
+  function detectLiteralDefaultValueCycle(
+    inputObj: GraphQLInputObjectType,
+    defaultValue: ConstValueNode,
+  ): void {
+    // If the value is a List, recursively check each entry for a cycle.
+    // Otherwise, only object values can contain a cycle.
+    if (defaultValue.kind === Kind.LIST) {
+      for (const itemLiteral of defaultValue.values) {
+        detectLiteralDefaultValueCycle(inputObj, itemLiteral);
+      }
+      return;
+    } else if (defaultValue.kind !== Kind.OBJECT) {
+      return;
+    }
+
+    // Check each defined field for a cycle.
+    const fieldNodes = keyMap(defaultValue.fields, (field) => field.name.value);
+    for (const field of Object.values(inputObj.getFields())) {
+      const fieldType = getNamedType(field.type);
+
+      // Only input object type fields can result in a cycle.
+      if (!isInputObjectType(fieldType)) {
+        continue;
+      }
+
+      // If the provided value has this field defined, recursively check it for
+      // cycles instead of checking the field's default value.
+      if (hasOwnProperty(fieldNodes, field.name)) {
+        detectLiteralDefaultValueCycle(fieldType, fieldNodes[field.name].value);
+        continue;
+      }
+
+      detectFieldDefaultValueCycle(inputObj, field, fieldType);
```

Comment:
As above, edit for clarity:

```suggestion
      const namedFieldType = getNamedType(field.type);

      // Only input object type fields can result in a cycle.
      if (!isInputObjectType(namedFieldType)) {
        continue;
      }

      // If the provided value has this field defined, recursively check it for
      // cycles instead of checking the field's default value.
      if (hasOwnProperty(fieldNodes, field.name)) {
        detectLiteralDefaultValueCycle(namedFieldType, fieldNodes[field.name].value);
        continue;
      }

      detectFieldDefaultValueCycle(inputObj, field, namedFieldType);
```